File: apps/frontend/src/adapters/shared/report-builder.ts
import { invoke } from "./platform";

export type ReportType =
  | "net_worth"
  | "portfolio_summary"
  | "income"
  | "data_quality"
  | "tax_pack"
  | "monthly_wealth_letter"
  | "estate_binder"
  | "fee_report";
export type ReportRunStatus = "generated" | "exported";
export type FeeCategory =
  | "broker_fees"
  | "transaction_fees"
  | "platform_fees"
  | "advisory_fees"
  | "fund_expense_ratio_manual"
  | "insurance_ulip_charges"
  | "fx_fees"
  | "private_fund_fees"
  | "custody_admin_fees"
  | "other";
export type EstateBinderSection =
  | "accounts"
  | "assets"
  | "liabilities"
  | "property"
  | "insurance"
  | "pensions"
  | "private_investments"
  | "documents_manifest"
  | "entity_ownership"
  | "islamic_notes";

export interface GenerateReportRequest {
  reportType: ReportType;
  baseCurrency: string;
  periodMonth?: string | null;
  includedSections?: EstateBinderSection[] | null;
}

export interface ManualFeeEntryInput {
  feeDate: string;
  category: FeeCategory;
  amount: string;
  currency: string;
  accountId?: string | null;
  assetId?: string | null;
  sourceCitationId?: string | null;
  notes?: string | null;
}

export interface ManualFeeEntry extends ManualFeeEntryInput {
  id: string;
  createdAt: string;
  updatedAt: string;
}

export interface FeeCategoryTotal {
  category: FeeCategory;
  amount: string;
  currency: string;
}

export interface FeeCurrencyTotal {
  amount: string;
  currency: string;
}

export interface FeeSpikeAlert {
  currency: string;
  currentPeriodTotal: string;
  priorAverage: string;
  multiple: string;
}

export interface FeeIntelligenceSummary {
  periodMonth: string;
  totals: FeeCurrencyTotal[];
  categoryTotals: FeeCategoryTotal[];
  spike: FeeSpikeAlert | null;
  missingFeesState: boolean;
}

export type ConcentrationDimension =
  | "asset"
  | "account_custodian"
  | "currency"
  | "sector_taxonomy"
  | "country_taxonomy"
  | "asset_class"
  | "income_source"
  | "manual_stale"
  | "private_illiquid"
  | "shariah_unknown"
  | "document_uncited";

export interface ConcentrationExposure {
  dimension: ConcentrationDimension;
  label: string;
  amount: string;
  currency: string;
  percent: string;
  sourceCount: number;
}

export interface ConcentrationFinding {
  dimension: ConcentrationDimension;
  label: string;
  message: string;
  amount: string;
  currency: string;
  percent: string;
  thresholdPercent: string;
}

export interface ConcentrationFragilitySummary {
  asOfDate: string;
  baseCurrency: string;
  totalWealth: string;
  exposures: ConcentrationExposure[];
  findings: ConcentrationFinding[];
  emptyState: boolean;
  islamicModeEnabled: boolean;
  taxonomyState: string;
}

export interface ReportLine {
  id: string;
  sectionId: string;
  label: string;
  amount: string | null;
  currency: string | null;
  valueText: string | null;
  sourceCitationId: string | null;
  metadataJson: string | null;
}

export interface ReportSection {
  id: string;
  reportRunId: string;
  title: string;
  sectionOrder: number;
  metadataJson: string | null;
  lines: ReportLine[];
}

export interface ReportRun {
  id: string;
  reportType: ReportType;
  baseCurrency: string;
  status: ReportRunStatus;
  createdAt: string;
  completedAt: string | null;
  sections: ReportSection[];
  disclaimer: string;
}

export interface ReportExportBundle {
  fileName: string;
  mimeType: string;
  bytes: number[];
}

export function generateReport(request: GenerateReportRequest): Promise<ReportRun> {
  return invoke<ReportRun>("generate_report", { request });
}

export function getReportRun(reportRunId: string): Promise<ReportRun | null> {
  return invoke<ReportRun | null>("get_report_run", { reportRunId });
}

export function exportReport(reportRunId: string): Promise<ReportExportBundle> {
  return invoke<ReportExportBundle>("export_report", { reportRunId });
}

export function addManualFeeEntry(input: ManualFeeEntryInput): Promise<ManualFeeEntry> {
  return invoke<ManualFeeEntry>("add_manual_fee_entry", { input });
}

export function getFeeIntelligenceSummary(
  periodMonth?: string | null,
): Promise<FeeIntelligenceSummary> {
  return invoke<FeeIntelligenceSummary>("get_fee_intelligence_summary", { periodMonth });
}

export function getConcentrationFragilitySummary(
  baseCurrency: string,
): Promise<ConcentrationFragilitySummary> {
  return invoke<ConcentrationFragilitySummary>("get_concentration_fragility_summary", {
    baseCurrency,
  });
}

File: apps/tauri/src/commands/report_builder.rs
use std::sync::Arc;

use log::error;
use mizan_core::report_builder::{
    FeeIntelligenceSummary, GenerateReportRequest, ManualFeeEntry, ManualFeeEntryInput,
    ReportBuilderRepositoryTrait, ReportExportBundle, ReportRun,
};
use tauri::State;

use crate::context::ServiceContext;

#[tauri::command]
pub async fn generate_report(
    request: GenerateReportRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ReportRun, String> {
    state
        .report_builder_repository()
        .generate_report(request)
        .await
        .map_err(command_error("generate_report"))
}

#[tauri::command]
pub async fn get_report_run(
    report_run_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Option<ReportRun>, String> {
    state
        .report_builder_repository()
        .get_report_run(&report_run_id)
        .map_err(command_error("get_report_run"))
}

#[tauri::command]
pub async fn export_report(
    report_run_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ReportExportBundle, String> {
    state
        .report_builder_repository()
        .export_report(&report_run_id)
        .map_err(command_error("export_report"))
}

#[tauri::command]
pub async fn add_manual_fee_entry(
    input: ManualFeeEntryInput,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ManualFeeEntry, String> {
    state
        .report_builder_repository()
        .add_manual_fee_entry(input)
        .await
        .map_err(command_error("add_manual_fee_entry"))
}

#[tauri::command]
pub async fn get_fee_intelligence_summary(
    period_month: Option<String>,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<FeeIntelligenceSummary, String> {
    state
        .report_builder_repository()
        .get_fee_intelligence_summary(period_month)
        .map_err(command_error("get_fee_intelligence_summary"))
}

fn command_error(command: &'static str) -> impl FnOnce(mizan_core::Error) -> String {
    move |err| {
        error!("{command} failed: {err}");
        format!("{command} failed: {err}")
    }
}
